### DISCUSSION TITLE
How does ThenBy work with Comparer?

### DISCUSSION THREAD

**Discussion by Morgfrimen** [dotnet/runtime#71343] [Q&A] [answered]
Good afternoon, colleagues. I have a question.

Program class:
```cs
var list = new List<SiteMapModel>()
{
    new() { Level = 0, Name = "A0", Publish = DateTime.UtcNow },
    new() { Level = 1, Name = "A1", Publish = DateTime.UtcNow.AddDays(5) },
    new() { Level = 1, Name = "A11", Publish = DateTime.UtcNow.AddDays(4) },
    new() { Level = 1, Name = "B1", Publish = DateTime.UtcNow.AddDays(6) },
    new() { Level = 1, Name = "C1", Publish = DateTime.UtcNow.AddDays(4) },
    new() { Level = 0, Name = "B0", Publish = DateTime.UtcNow.AddDays(1) },
    new() { Level = 0, Name = "C0", Publish = DateTime.Now },
    new() { Level = 2, Name = "A2", Publish = DateTime.UtcNow.AddDays(5) }
};

var comparerList = list.OrderBy(x => x.Name).ToList().OrderBy(x => x, new SiteMapComparer());

foreach (var model in comparerList)
{
    Console.WriteLine($"{new string(' ', 2 * model.Level)}{model.Name}(level:{model.Level})");
}
```
Model:
```cs
public sealed class SiteMapModel
{
    public string Name { get; set; }
    public int Level { get; set; }
    public DateTime Publish { get; set; }
}
```
IComparer object:
```cs
public sealed class SiteMapComparer : IComparer<SiteMapModel>
{
    public int Compare(SiteMapModel? x, SiteMapModel? y)
    {
        if (ReferenceEquals(x, y))
            return 0;

        if (ReferenceEquals(null, y))
            return 1;

        if (ReferenceEquals(null, x))
            return -1;

        var nameComparison = string.Compare(x.Name, y.Name, StringComparison.Ordinal);
        var dateCompare = DateTime.Compare(x.Publish, y.Publish);

        if (x.Level != y.Level)
        {
            return nameComparison;
        }

        if (x.Level == y.Level)
        {
            return dateCompare;
        }

        return default;
    }
}
```

If writing is not beautiful:
```cs
var comparerList = list.OrderBy(x => x.Name).ToList().OrderBy(x => x, new SiteMapComparer());
```
Sorting works as I want!
But
```cs
var comparerList = list.OrderBy(x => x.Name).ThenBy(x => x, new SiteMapComparer());
```
I don't get into Comparer and sorting doesn't work as it should!

Why is this happening?

**Comment by heathbm** [comment, marked as answer]
First, in your initial chained `OrderBy`, you can drop the `ToList()`.

`var comparerList = list.OrderBy(x => x.Name).ToList().OrderBy(x => x, new SiteMapComparer());`
to
`var comparerList = list.OrderBy(x => x.Name).OrderBy(x => x, new SiteMapComparer());`

To answer the `ThenBy` question:

Essentially, `ThenBy` does not perform a complete `OrderBy` again, otherwise, it would just duplicate `OrderBy()`.
`OrderBy` will order the list by a key (`x.Name` in this case).
`ThenBy` allows you to apply subsequent ordering on the items belonging to the same key.
Since your provided list has no duplicate `Name` keys, the `SiteMapComparer` will never be called as there is only 1 item to sort, so it is skipped.

**Reply by Clockwork-Muse** [reply]
> First, in your initial chained `OrderBy`, you can drop the `ToList()`.

Officially, per the documentation:
> Because [IOrderedEnumerable<TElement>](https://docs.microsoft.com/en-us/dotnet/api/system.linq.iorderedenumerable-1?view=net-6.0) inherits from [IEnumerable<T>](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.ienumerable-1?view=net-6.0), you can call [OrderBy](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.orderby?view=net-6.0) or [OrderByDescending](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.orderbydescending?view=net-6.0) on the results of a call to [OrderBy](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.orderby?view=net-6.0), [OrderByDescending](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.orderbydescending?view=net-6.0), [ThenBy](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.thenby?view=net-6.0) or [ThenByDescending](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.thenbydescending?view=net-6.0). **Doing this introduces a new primary ordering that ignores the previously established ordering.**

... this means that the two legally may have distinct behaviors, because by capturing the list the collection is guaranteed to be ordered by the first key, while if it is dropped the results of the first `OrderBy` call could be ignored completely.
(It seems to produce the expected results currently, although I'm not sure whether this is intended or merely an implementation detail)
